Lib/ensurepip/_structs.py

```diff
@@ -0,0 +1,251 @@
+    def from_remote_dist(
+            cls,
+            remote_distribution_pkg: _RemoteDistributionPackage,
+            /,
+    ) -> BundledDistributionPackage:
+        """Create a replacement package from its remote counterpart."""
+        return cls.from_project_spec(
+            remote_distribution_pkg.project_name,
+            remote_distribution_pkg.project_version,
+        )
+
+    def as_pathlib_ctx(self) -> AbstractContextManager[Path]:
+        """Make a context manager exposing a :mod:`pathlib` instance."""
+        return _traversable_to_pathlib_ctx(self.wheel_path)
+
+
+@dataclass(frozen=True)
+class ReplacementDistributionPackage:
+    """Structure representing a wheel under ``WHEEL_PKG_DIR``."""
+
+    project_name: str
+    """PyPI project name."""
+
+    project_version: str
+    """PyPI project version."""
+
+    wheel_name: str
+    """Wheel package file name."""
+
+    wheel_path: Path
+    """Wheel package file path."""
+
+    @classmethod
+    def from_distribution_package_name(
+            cls,
+            dist_pkg_name: str,
+            /,
+    ) -> ReplacementDistributionPackage:
+        """Look up a replacement package from name.
+
+        :raises LookupError: If ``WHEEL_PKG_DIR`` is not set or the package
+                             wheel is nowhere to be found.
+
+        """
+        wheel_file_name_prefix = f'{dist_pkg_name !s}-'
+
+        wheel_pkg_dir = _get_wheel_pkg_dir_from_sysconfig()
+        dist_matching_wheels = wheel_pkg_dir.glob(
+            f'{wheel_file_name_prefix !s}*.whl',
+        )
+
+        try:
+            first_matching_dist_wheel = sorted(dist_matching_wheels)[0]
+        except IndexError as index_err:
+            raise LookupError(
+                '`WHEEL_PKG_DIR` does not contain any wheel files '
+                f'for `{dist_pkg_name !s}`.',
+            ) from index_err
+
+        wheel_name = first_matching_dist_wheel.name
+        dist_pkg_version = (
+            # Extract '21.2.4' from 'pip-21.2.4-py3-none-any.whl'
+            wheel_name.
+            removeprefix(wheel_file_name_prefix).
+            partition('-')[0]
+        )
+
+        return cls(
+            dist_pkg_name, dist_pkg_version,
+            wheel_name, first_matching_dist_wheel,
+        )
+
+    @classmethod
+    def from_remote_dist(
+            cls,
+            remote_distribution_pkg: _RemoteDistributionPackage,
+            /,
+    ) -> ReplacementDistributionPackage:
+        """Create a replacement package from its remote counterpart."""
+        return cls.from_distribution_package_name(
+            remote_distribution_pkg.project_name,
+        )
+
+    def as_pathlib_ctx(self) -> AbstractContextManager[Path]:
+        """Make a context manager exposing a :mod:`pathlib` instance."""
+        return nullcontext(self.wheel_path)
+
+
+PIP_REMOTE_DIST = _RemoteDistributionPackage(
+    'pip',
+    '23.2.1',
+    '7ccf472345f20d35bdc9d1841ff5f313260c2c33fe417f48c30ac46cccabf5be',
```

Comment:
Thanks. So the pip release manager would need to calculate that sha256 value? That should be OK, although while I have tools on Windows to do this, I don't know if everyone does (I assume Unix has a command for this, although I've never checked).

(Actually, I just saw that you mentioned it can be copied from PyPI, so that's sufficient).